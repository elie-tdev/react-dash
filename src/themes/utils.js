export const opacity = (hex, opacity) => {
  var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  if (!result) {
    throw new Error('Invalid hex', hex);
  }
  return `rgba(${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(
    result[3],
    16,
  )}, ${opacity})`;
};

export const gradient = ({ type = 'linear', colors, background }) =>
  `${type}-gradient(${colors.join(', ')}), ${background};`;
