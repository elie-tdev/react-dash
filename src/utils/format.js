const formatter = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'USD',
  minimumFractionDigits: 2,
});

const parseFloat = (str, val = 2) => {
  str = str && str.toString();
  str = str && str.indexOf('.') > -1 ? str.slice(0, str.indexOf('.') + val + 1) : str;
  return Number(str);
};

export const formatDollar = number => formatter.format(parseFloat(number, 2));
