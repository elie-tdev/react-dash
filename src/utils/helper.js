const maskAccountNumber = accNumber => {
  return accNumber && <>&bull;&bull;&bull;&bull; {accNumber.slice(accNumber.length - 4)}</>;
};

export default maskAccountNumber;
