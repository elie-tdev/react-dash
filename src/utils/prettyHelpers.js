export const prettyCurrency = (amount, currency = 'USD') => {
  if (amount == null) {
    return false;
  }
  if (amount < 0) {
    amount = 0;
  }
  return amount.toLocaleString('en', { style: 'currency', currency });
};

export const prettyDateTime = time => {
  if (time == null) {
    return false;
  }
  let d = new Date(time);
  return (
    d.toLocaleDateString('en', {
      month: '2-digit',
      day: '2-digit',
      year: '2-digit',
    }) +
    ' at ' +
    d.toLocaleTimeString('en', { hour: '2-digit', minute: '2-digit' }).replace(/^0(?:0:0?)?/, '')
  );
};

export const prettyDate = time => new Date(time).toLocaleDateString('en', {});

export const prettyTime = time =>
  new Date(time).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });

export const prettyHour = time =>
  new Date(time)
    .toLocaleTimeString('en', { hour: '2-digit', minute: '2-digit' })
    .replace(/^0(?:0:0?)?/, '');

export const prettyPhone = number =>
  number.substr(2).replace(/(\d{1,2})(\d{1})?(\d{1,3})?(\d{1,4})?/, function (_, p1, p2, p3, p4) {
    let output = '';
    if (p1) output = `${p1}`;
    if (p2) output += `${p2}-`;
    if (p3) output += `${p3}-`;
    if (p4) output += `${p4}`;
    return output;
  });

export const minimumDays = (date, numDays) => {
  const d = new Date(date);
  const totalNumDays = daysIntoYear(d) + numDays;
  return Boolean(daysIntoYear(d) >= totalNumDays || totalNumDays <= 365);
};

export const daysIntoYear = date => {
  return (
    (Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()) -
      Date.UTC(date.getFullYear(), 0, 0)) /
    24 /
    60 /
    60 /
    1000
  );
};

export const cleanAmountRegex = value => {
  return value.replace(/[^0-9]/g, '');
};

export const cleanAmount = value => {
  const pos = value.indexOf('.');
  let result;
  if (pos !== -1) {
    const part1 = value.substr(0, pos);
    const part2 = value.substr(pos + 1);
    result = cleanAmountRegex(part1) + '.' + cleanAmountRegex(part2);
  } else {
    result = cleanAmountRegex(value);
  }
  return result;
};
