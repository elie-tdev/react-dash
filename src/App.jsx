import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { CssBaseline, CircularProgress } from '@material-ui/core';
import React, { useEffect, useMemo, useState } from 'react';
import { ApolloClient, ApolloProvider, createHttpLink, InMemoryCache } from '@apollo/client';
import { useAuth0, Auth0Provider } from '@auth0/auth0-react';
import { useLazyQuery } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import { useHistory, useRouteMatch } from 'react-router-dom';

import {
  apiUrl,
  REACT_APP_AUTH0_DOMAIN,
  REACT_APP_AUTH0_CLIENT_ID,
  REACT_APP_AUTH0_AUDIENCE,
} from './properties';

import { queries } from '@/gql';
import { LightTheme } from '@/themes';
import { MenuContextProvider } from '@/contexts';
import {
  LoanOverviewScreen,
  PaymentHistoryScreen,
  StatementsScreen,
  PreferencesScreen,
  PaymentAccountsScreen,
  HelpCenterScreen,
  ChangePasswordScreen,
  ChangeAddressScreen,
  BalanceScreen,
} from '@/screens';
import { MobileMenu, CenteredScreenContent, Screen } from '@/components';
import {
  OVERVIEW_ROUTE,
  BALANCE_ROUTE,
  PAYMENTS_ROUTE,
  STATEMENTS_ROUTE,
  PREFERENCES_ROUTE,
  PAYMENT_ACCOUNTS_ROUTE,
  HELP_CENTER_ROUTE,
  CHANGE_PASSWORD_ROUTE,
  CHANGE_ADDRESS_ROUTE,
  LOGOUT_ROUTE,
} from '@/constants/routes';
import { useAuthUser } from '@/customHooks';

const httpLink = createHttpLink({
  uri: `${apiUrl}/graphql`,
});

const App = () => {
  const history = useHistory();
  const match = useRouteMatch('/dashboard/loans/:loanNumber');
  const { contactId } = useAuthUser();

  const [fetchInitialUserData, { data, loading, error }] = useLazyQuery(queries.InitializeUser, {
    variables: { contact_id: contactId },
  });

  useEffect(() => {
    if (contactId) {
      fetchInitialUserData();
    }
  }, [contactId]);

  const loans = useMemo(() => data?.Contacts_GetLoans?.payload?.data || [], [data]);

  const loanNumber = match?.params?.loanNumber;

  useEffect(() => {
    if (
      loans?.length > 0 &&
      (!loanNumber || (loanNumber && !loans?.find(loan => loan.Acctrefno === loanNumber)))
    ) {
      history.push(`/dashboard/loans/${loans[0].Acctrefno}`);
    }
  }, [loans, loanNumber]);

  if (loading) {
    return (
      <Screen>
        <CenteredScreenContent minHeight="75vh">
          <CircularProgress color="primary" />
        </CenteredScreenContent>
      </Screen>
    );
  }

  if (error) {
    return (
      <Screen>
        <CenteredScreenContent minHeight="75vh">
          <div>{error.message}</div>
        </CenteredScreenContent>
      </Screen>
    );
  }

  return (
    <>
      <MobileMenu />
      <Switch>
        <Route exact path={OVERVIEW_ROUTE}>
          <LoanOverviewScreen />
        </Route>
        <Route exact path={BALANCE_ROUTE}>
          <BalanceScreen />
        </Route>
        <Route path={PAYMENTS_ROUTE}>
          <PaymentHistoryScreen />
        </Route>
        <Route path={STATEMENTS_ROUTE}>
          <StatementsScreen />
        </Route>
        <Route path={PAYMENT_ACCOUNTS_ROUTE}>
          <PaymentAccountsScreen />
        </Route>
        <Route path={CHANGE_PASSWORD_ROUTE}>
          <ChangePasswordScreen />
        </Route>
        <Route path={CHANGE_ADDRESS_ROUTE}>
          <ChangeAddressScreen />
        </Route>
        <Route path={PREFERENCES_ROUTE}>
          <PreferencesScreen />
        </Route>
        <Route path={HELP_CENTER_ROUTE}>
          <HelpCenterScreen />
        </Route>
      </Switch>
    </>
  );
};

const ApolloContainer = () => {
  const { getAccessTokenSilently, isAuthenticated } = useAuth0();
  const [token, setToken] = useState('');

  useEffect(() => {
    const getToken = async () => {
      const token = await getAccessTokenSilently();
      setToken(token);
    };
    if (!token && isAuthenticated) {
      getToken();
    }
  }, [isAuthenticated]);

  if (!token) {
    return (
      <CenteredScreenContent>
        <CircularProgress color="primary" />
      </CenteredScreenContent>
    );
  }

  const authLink = setContext((_, { headers }) => ({
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : '',
    },
  }));

  const client = new ApolloClient({
    link: authLink.concat(httpLink),
    cache: new InMemoryCache(),
  });

  return (
    <ApolloProvider client={client}>
      <ThemeProvider theme={LightTheme}>
        <MenuContextProvider>
          <CssBaseline />
          <App />
        </MenuContextProvider>
      </ThemeProvider>
    </ApolloProvider>
  );
};

const Auth0Container = () => {
  const { logout } = useAuth0();
  const history = useHistory();
  const { isAuthenticated, loginWithRedirect, isLoading, error } = useAuth0();

  useEffect(() => {
    if (history.location.pathname === LOGOUT_ROUTE) {
      logout({ returnTo: `${window.location.origin}/dashboard/loans/` });
    }
  }, []);

  useEffect(() => {
    if (!isAuthenticated && !isLoading) {
      loginWithRedirect();
    }
  }, [isAuthenticated, isLoading]);

  if (isLoading) {
    return (
      <CenteredScreenContent>
        <CircularProgress color="primary" />
      </CenteredScreenContent>
    );
  }

  if (error) {
    return (
      <CenteredScreenContent>
        <div>{error.message}</div>
      </CenteredScreenContent>
    );
  }

  return <ApolloContainer />;
};

const RouterContainer = () => (
  <Router>
    <Auth0Provider
      domain={REACT_APP_AUTH0_DOMAIN}
      clientId={REACT_APP_AUTH0_CLIENT_ID}
      redirectUri={`${window.location.origin}/dashboard/loans/`}
      audience={`https://${REACT_APP_AUTH0_AUDIENCE}`}
      scope={'openid profile email'}
    >
      <Auth0Container />
    </Auth0Provider>
  </Router>
);

export default RouterContainer;
