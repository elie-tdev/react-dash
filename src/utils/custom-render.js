import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { render } from '@testing-library/react';
import { CssBaseline } from '@material-ui/core';
import { BrowserRouter as Router } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { MockedProvider } from '@apollo/client/testing';

import { MenuContextProvider } from '@/contexts';
import i18n from '@/translations';
import { LightTheme } from '@/themes';

const wrapper = ({ mocks = [], menuContextProviderProps = {} }) => {
  const Wrapper = ({ children }) => {
    return (
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <ThemeProvider theme={LightTheme}>
            <MenuContextProvider {...menuContextProviderProps}>
              <CssBaseline />
              <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
            </MenuContextProvider>
          </ThemeProvider>
        </MockedProvider>
      </Router>
    );
  };

  return Wrapper;
};

const customRender = (ui, { mocks, menuContextProviderProps, ...restOptions } = {}) =>
  render(ui, { wrapper: wrapper({ mocks, menuContextProviderProps }), ...restOptions });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
