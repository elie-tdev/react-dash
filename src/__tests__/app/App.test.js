import React from 'react';

import App from '@/App';
import { render } from '@/utils/custom-render';

jest.mock('@auth0/auth0-react', () => ({
  ...jest.requireActual('@auth0/auth0-react'),
  useAuth0: () => ({
    user: {
      [`https://${process.env.REACT_APP_AUTH0_AUDIENCE}/metadata`]: {
        : {
          cifno: '206048',
        },
      },
    },
    loginWithRedirect: () => null,
    getAccessTokenSilently: () => null,
  }),
}));

describe('App Component', () => {
  test('renders log in', () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
  });
});
