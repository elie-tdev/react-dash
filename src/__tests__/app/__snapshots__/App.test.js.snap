// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders log in 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-4 makeStyles-contentBox-2 makeStyles-contentBox-3"
  >
    <div
      class="MuiBox-root MuiBox-root-5 makeStyles-childrenBox-1"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
